demos/device-shadow-for-aws-iot-embedded-sdk/shadow_demo_main.c

```diff
@@ -0,0 +1,679 @@
+            }
+            else
+            {
+                LogInfo( ( "Other message type:%d !!\n\n", messageType ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Shadow_MatchTopic parse failed:%s !!\n\n", ( const char * ) pxDeserializedInfo->pPublishInfo->pTopicName ) );
+        }
+    }
+    else
+    {
+        vHandleOtherIncomingPacket( pxPacketInfo, usPacketIdentifier );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Entry point of shadow demo.
+ *
+ * This main function demonstrates how to use the macros provided by the
+ * Device Shadow library to assemble strings for the MQTT topics defined
+ * by AWS IoT Device Shadow. It uses these macros for topics to subscribe
+ * to:
+ * - SHADOW_TOPIC_STRING_UPDATE_DELTA for "$aws/things/thingName/shadow/update/delta"
+ * - SHADOW_TOPIC_STRING_UPDATE_ACCEPTED for "$aws/things/thingName/shadow/update/accepted"
+ * - SHADOW_TOPIC_STRING_UPDATE_REJECTED for "$aws/things/thingName/shadow/update/rejected"
+ *
+ * It also uses these macros for topics to publish to:
+ * - SHADOW_TOPIC_STIRNG_DELETE for "$aws/things/thingName/shadow/delete"
+ * - SHADOW_TOPIC_STRING_UPDATE for "$aws/things/thingName/shadow/update"
+ *
+ * The helper functions this demo uses for MQTT operations have internal
+ * loops to process incoming messages. Those are not the focus of this demo
+ * and therefor, are placed in a separate file shadow_demo_helpers.c.
+ */
+int RunDeviceShadowDemo( bool awsIotMqttMode,
+                         const char * pIdentifier,
+                         void * pNetworkServerInfo,
+                         void * pNetworkCredentialInfo,
+                         const void * pNetworkInterface )
+{
+    int returnStatus = EXIT_SUCCESS;
+
+    /* A buffer containing the update document. It has static duration to prevent
+     * it from being placed on the call stack. */
+    static char pcUpdateDocument[ SHADOW_REPORTED_JSON_LENGTH + 1 ] = { 0 };
+
+    /* Remove compiler warnings about unused parameters. */
+    ( void ) awsIotMqttMode;
+    ( void ) pIdentifier;
+    ( void ) pNetworkServerInfo;
+    ( void ) pNetworkCredentialInfo;
+    ( void ) pNetworkInterface;
+
+    returnStatus = EstablishMqttSession( prvEventCallback );
+
+    if( returnStatus == EXIT_FAILURE )
+    {
+        /* Log error to indicate connection failure. */
+        LogError( ( "Failed to connect to MQTT broker." ) );
+    }
+    else
+    {
+        /* First of all, try to delete any Shadow document in the cloud. */
+        returnStatus = PublishToTopic( SHADOW_TOPIC_STRING_DELETE( THING_NAME ),
+                                       SHADOW_TOPIC_LENGTH_DELETE( THING_NAME_LENGTH ),
+                                       pcUpdateDocument,
+                                       0U );
+
+        /* Successfully connect to MQTT broker, the next step is
```

Comment:
You already connected at line 539, please delete mentioning it here.

Reply:
thanks!